src/main/groovy/de/aaschmid/gradle/plugins/cpd/Cpd.groovy

```diff
@@ -174,57 +172,49 @@ class Cpd extends SourceTask implements VerificationTask, Reporting<CpdReports>
 
-        // use getter to access properties that they are resolved correctly using conventionMapping
-        CpdReporter reporter = new CpdReporter(this)
-        CpdExecutor executor = new CpdExecutor(this)
+        Properties languageProperties = new Properties();
 
-        List<Match> matches = executor.run()
-        reporter.generate(matches)
-        logResult(matches)
-    }
+        if (isIgnoreLiterals()) {
+            languageProperties.setProperty(Tokenizer.IGNORE_LITERALS, "true");
+        }
+        if (isIgnoreIdentifiers()) {
+            languageProperties.setProperty(Tokenizer.IGNORE_IDENTIFIERS, "true");
+        }
+        if (isIgnoreAnnotations()) {
+            languageProperties.setProperty(Tokenizer.IGNORE_ANNOTATIONS, "true");
+        }
+        languageProperties.setProperty(Tokenizer.OPTION_SKIP_BLOCKS, Boolean.toString(isSkipBlocks()));
+        languageProperties.setProperty(Tokenizer.OPTION_SKIP_BLOCKS_PATTERN, getSkipBlocksPattern());
+
+        def enabledReports = getReports().getEnabled()
 
-    private void logResult(List<Match> matches) {
-        if (matches.isEmpty()) {
-            if (logger.isInfoEnabled()) {
-                logger.info('No duplicates over {} tokens found.', getMinimumTokenCount())
-            }
-
-        } else {
-            String message = "CPD found duplicate code.";
-            SingleFileReport report = reports.getFirstEnabled();
-            if (report != null) {
-                String reportUrl = asClickableFileUrl(report.getDestination());
-                message += " See the report at ${reportUrl}";
-            }
-            if (getIgnoreFailures()) {
-                if (logger.isWarnEnabled()) {
-                    logger.warn(message);
-                }
-            } else {
-                throw new GradleException(message);
-            }
+        if (enabledReports.isEmpty()) {
+            throw new InvalidUserDataException("All reports for task '$name' are disabled.")
         }
-    }
 
-    private String asClickableFileUrl(File path) {
-        // File.toURI().toString() leads to an URL like this on Mac: file:/reports/index.html
-        // This URL is not recognized by the Mac console (too few leading slashes). We solve
-        // this be creating an URI with an empty authority.
-        try {
-            return new URI("file", "", path.toURI().getPath(), null, null).toString();
-        } catch (URISyntaxException e) {
-            throw UndeclaredThrowableException(e);
+        workerExecutor.submit(CpdAction) { WorkerConfiguration config ->
+            config.classpath(getPmdClasspath())
+            config.setParams(
+                    getEncoding(),
+                    getMinimumTokenCount(),
+                    getLanguage(),
+                    languageProperties,
+                    isSkipLexicalErrors(),
+                    isSkipDuplicateFiles(),
+                    new HashSet<>(getSource().getFiles()),
```

Comment:
It would be possible, but what would be the benefit?

IMHO an unmodifiable set is only interesting for sets which are used for a long time and which are accessed/shared by multiple components.

Reply:
Hm ... I thought more about the performance as copying the set is more expensive, even more on hugh projects ...